### DISCUSSION TITLE
Notation sign/verify arbitrary file

### DISCUSSION THREAD

**Discussion by ghost** [notaryproject/notation#767] [Ideas]
UX proposal for `sign` in file system:
```shell
# Prerequisites:
# A default signing key is configured using CLI "notation key"

# Use boolean flag "--file" to enable signing a file
# The generated signature is saved at the same dir as the target file with name
# <target_file>.sig
notation sign --file <target_file_path>:<mediaType>

# Use boolean flag "--file" to enable signing a file
# Use flag "--signature" to specify path where the generated signature is stored
notation sign --file <target_file_path>:<mediaType> --signature <signature_path>
```
-----------------------------------------------------------------------------------------------------------
UX proposal for `verify` in file system:
```shell
# Prerequisites: Both target and signature files are stored in file system

# Use boolean flag "--file" to enable verifying a file
# Use flag "--signature" to specify path where the signature is stored
# The global trust policy is used by default
notation verify --file <target_file_path>:<mediaType> --signature <signature_path>

# Use boolean flag "--file" to enable verifying a file
# Use flag "--signature" to specify path where the signature is stored
# Trust policy with scope "example/myPolicy" is used, if it does not exist, the global trust policy is used
notation verify --file <target_file_path>:<mediaType> --signature <signature_path> --scope <example/myPolicy>
```
UX proposal for `verify` file stored in OCI-compliant registry:
```shell
# Prerequisites: Signatures are stored in the registry referencing the file artifact

# Use boolean flag "--file" to enable verifying a file
notation verify --file <full_URI_reference>
```

The above CLI design requires following changes in the [specifications](https://github.com/notaryproject/specifications) repo:
1. Signature payload:
    ```bash
    {
      "targetArtifact": {
        "mediaType": "sbom/example",             // can be defined by user, but we need a default value
        "digest": "sha256:abcdef",               // digest of the file content
        "size": 32654                            // size of the file content
      }
    }
    ```
2. Signature storage in file system
3. Trust policy: for files and signatures stored in file system, the `registryScope` term needs to be redefined.
4. Any sign and verify workflow changes involved in this new process
@yizha1 @FeynmanZhou

Based on the above design, there is a major discussion started by @priteshbandi:
notation should sign/verify arbitrary `data` instead of just `files`. For example, data from stdin (we need a clearer scenario on this one).
Here's my understanding: the `--file` flag is introduced based on our discussion when we brought in the `--oci-layout` flag several months ago: https://github.com/notaryproject/notation/pull/601#discussion_r1150810148. We are following the practice of `one type per flag` (this is also what ORAS is following). Thus, to sign/verify data from stdin, we'd have another flag such as `--stdin`. Another example is, I'd like to sign/verify a `descriptor` using notation. In this case, a `--descriptor` flag is necessary.
And of course, my understanding might not perfectly align with what our users are expecting. Therefore, we need more people's inputs to discuss on this part. @sajayantony @shizhMSFT @toddysm @yizha1 @FeynmanZhou

**Comment by shizhMSFT** [comment]
Is `--file` a flag of `--file bool` or `--file string`?

**Reply by ghost** [reply]
It's a --file bool. When --file flag is set, `<reference>` of the command becomes `<target_file_path>`. Updated the original post.

**Comment by yizha1** [comment]
@priteshbandi @Two-Hearts @shizhMSFT @FeynmanZhou I would suggest the following process for feature https://github.com/notaryproject/notation/issues/741. What do you think? Of cause, we can continue discussing specifications and UX here before we finalize it in respective repo.

1. Finalize new specs under `specifications` repo, see two issues
- https://github.com/notaryproject/specifications/issues/275 for scenario 1: sign and verify arbitrary blobs
- https://github.com/notaryproject/specifications/issues/276 for scenario 2: distribute arbitrary blobs and signatures via OCI compliant registries and verify signatures in registries.
2. Finalize CLI specs under `notation` repo for CLI UX
3. Implementation

Note: We may complete scenario 1 first then scenario 2.

**Comment by priteshbandi** [comment]
IMO we should start with requirements/usecases then work on user experience and once we have we UX finalized, we can make relevant changes to NotaryProject specifications. At high level I see following usecases

### Requirements
1. User MUST be able to sign and verify arbitrary content
2. For verification, the user MUST be able to leverage existing fine-grain controls(expiry, revocation, authenticity/logging, enforce) offered by the trust policy.
    1. The user SHOULD be able to configure default values for the controls which will by default apply to all verify operations.
    2. User SHOULD be able to specify a set of predefined controls (different from the default) for a given verify operation 
4. For signing, a user SHOULD be able to sign a precalculated digest. This is different because here notation won't recalculate digest instead sign whatever the user provided. (This will also help us confirm the extensibility for different kind of artifact)
5. For signing and verification, the user SHOULD be able to pass content as std-in or can be read from a file. std-in allows for better chaining/piping between tools


PS: we should go through comment on https://github.com/notaryproject/notation/pull/765 to confirm that we didnt miss any requirement

**Reply by ghost** [reply]
My understanding on `requirement 3`, signing a `digest` is a subset of signing a `descriptor`. 
A `descriptor` contains, quote from @shizhMSFT:
Digest ensures integrity
Size is for preventing endless data attack
Media type is for preventing content mismatch

**Reply by priteshbandi** [reply]
Yes but when we sign a descriptor, we rehash it and following scenario wont be possible. 

#### Sign
``` SomeFile --> DigestCalculatedUsingOpenSSL --> Sign Hash with Notation (`notation sign --digest sha256 ${digest}) ```

#### Verify
``` SomeFile --> Verify with notation (`notation verify --file ${pathToSomeFile}`) ```
Instead of above user will need to do following
``` SomeFile --> DigestCalculatedUsingOpenSSL --> Verify with notation (`notation verify --file ${pathToFileContaigningDigest}`) ```


One way to do above is using a specific flag while signing (lets say `--digest`) and then when forming descriptor to sign, the digest would be whatever user provided and size would be size of digest.

**Reply by shizhMSFT** [reply]
Digest of the file content only is not sufficient to provide security. We require `size` to prevent endless data attack and also media type for preventing content mismatch.

Here's the flow:
- **Sign**: `SomeFile` --> Calculate `descriptor` using `sha256sum` and `stat` with media type specified --> Sign `descriptor` using `notation sign`
- **Verify**: Verification can have two methods. Either is fine depending on the scenarios.
  1. `SomeFile` --> Calculate `descriptor` using `sha256sum` and `stat` with media type specified --> Verify `descriptor` using `notation verify`
  2. Verify signature and output the `descriptor` in the signature using `notation verify`. Check if `SomeFile` matches the `size` and `mediaType` in the `descriptor`. If not, we don't need to waste CPU time to compute digest of `SomeFile`. If the `size` matches, the digest of `SomeFile` should be computed and checked.

**Reply by shizhMSFT** [reply]
Let me elaborate more on the security provided by `size` and `mediaType`.

If the `descriptor` in the signature says the signed file content is of size `4 KiB` but the actual large file content is `32 TiB`, we should not bother computing the digest by reading the big file.

If the `descriptor` in the signature says the signed file content is a json file where `mediaType` is `application/json`, but the actual file is an image with `mediaType` of `image/png`. Even if the `size` and `digest` match, we should still reject the file.

**Reply by shizhMSFT** [reply]
For the mismatch vulnerability by not verifying `mediaType`, you may refer to [CVE-2021-41190](https://github.com/advisories/GHSA-mc8v-mgrf-8f4m).

**Reply by shizhMSFT** [reply]
A more clearer example would be this.

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 1470,
    "digest": "sha256:9c7a54a9a43cca047013b82af109fe963fde787f63f9e016fdc3384500c2823d"
  },
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 2457,
      "digest": "sha256:719385e32844401d57ecfd3eacab360bf551a1491c05b85806ed8f1b08d792f6"
    }
  ],
"architecture": "amd64",
  "config": {
    "Hostname": "",
    "Domainname": "",
    "User": "",
    "AttachStdin": false,
    "AttachStdout": false,
    "AttachStderr": false,
    "Tty": false,
    "OpenStdin": false,
    "StdinOnce": false,
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ],
    "Cmd": [
      "/hello"
    ],
    "Image": "sha256:62a15619037f3c4fb4e6ba9bd224cba3540e393a55dc52f6bebe212ca7b5e1a7",
    "Volumes": null,
    "WorkingDir": "",
    "Entrypoint": null,
    "OnBuild": null,
    "Labels": null
  },
  "container": "347ca68872ee924c4f9394b195dcadaf591d387a45d624225251efc6cb7a348e",
  "container_config": {
    "Hostname": "347ca68872ee",
    "Domainname": "",
    "User": "",
    "AttachStdin": false,
    "AttachStdout": false,
    "AttachStderr": false,
    "Tty": false,
    "OpenStdin": false,
    "StdinOnce": false,
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
    ],
    "Cmd": [
      "/bin/sh",
      "-c",
      "#(nop) ",
      "CMD [\"/hello\"]"
    ],
    "Image": "sha256:62a15619037f3c4fb4e6ba9bd224cba3540e393a55dc52f6bebe212ca7b5e1a7",
    "Volumes": null,
    "WorkingDir": "",
    "Entrypoint": null,
    "OnBuild": null,
    "Labels": {}
  },
  "created": "2023-05-04T17:37:03.872958712Z",
  "docker_version": "20.10.23",
  "history": [
    {
      "created": "2023-05-04T17:37:03.801840823Z",
      "created_by": "/bin/sh -c #(nop) COPY file:201f8f1849e89d53be9f6aa76937f5e209d745abfd15a8552fcf2ba45ab267f9 in / "
    },
    {
      "created": "2023-05-04T17:37:03.872958712Z",
      "created_by": "/bin/sh -c #(nop)  CMD [\"/hello\"]",
      "empty_layer": true
    }
  ],
  "os": "linux",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:01bb4fce3eb1b56b05adf99504dafd31907a5aadac736e36b27595c8b92f07f1"
    ]
  }
}
```
The above JSON file is a valid manifest of type `application/vnd.docker.distribution.manifest.v2+json` as well as a valid config of type `application/vnd.docker.container.image.v1+json`. The attacker can compose a valid config file and trick the user saying that it is a valid manifest since the size and digest exactly match the signature.

**Reply by priteshbandi** [reply]
We dont really need to understand/decode the data that we are signing or verifying. Is [CVE-2021-41190](https://github.com/advisories/GHSA-mc8v-mgrf-8f4m) applicable in that case?

**Reply by shizhMSFT** [reply]
In the above case, we are not decoding the data since the `mediaType` should be passed to `notation verify` by the user if the first verification method is used.

**Reply by gokarnm** [reply]
I assume this feature is for signing any kind of file on disk. Given a file, how will we deduce the media type during verification, to match with the media type in the signature. For e.g. given a image manifest, the verification workflow can interpret it as plain text, json or image manifest type. Am I missing something here?

**Reply by ghost** [reply]
@gokarnm Here's my understanding, instead of deducing to match the signature, a verifier should know what they are verifying (for example, this info can be provided by the signer). The UX for verify is:
`notation verify --file <target_file_path>:<mediaType> --signature <signature_path>`. 
If the signature payload's media type does not match with the verifier's expectation, then the verification would fail.

**Reply by priteshbandi** [reply]
> a verifier should know what they are verifying (for example, this info can be provided by the signer). The UX for verify is:
notation verify --file <target_file_path>:<mediaType> --signature <signature_path>.
If the signature payload's media type does not match with the verifier's expectation, then the verification would fail.

Why does a user need to know about mediaType to verify the integrity and authenticity of the data? For example, A user signed a file, Zduring the signing, they didn't provide media type (or if it's compulsory provided some value). Now during verification, if user  forgot mediatype. Shouldn't they be able to verify the integrity and authenticity of the data?

**Comment by yizha1** [comment]
@priteshbandi as one action from community meeting on 8/28/2023, could you share your thoughts on using `mediaType`? If we are aligned on the solution, I will create a PR to add a new spec for signing non-OCI artifacts.

**Comment by yanivpaz** [comment]
@toddysm is this solution suppose to also sign AWS lambdas?
